File: src/main/scala/demo/scala/LoanPattern.scala
package demo.scala
import java.io.{File, FileInputStream}

object LoanPattern {
  def readFile[T](f: File)(handler: FileInputStream => T): T = {
    val resource = new java.io.FileInputStream(f)
    try {
      handler(resource)
    } finally {
      resource.close()
    }
  }

  readFile(new java.io.File("test.txt")) { input =>
    var data = input.read
    while (data != -1) {
      print(data.toChar)
      data = input.read
    }
  }
}

File: src/main/java/demo/java/LoanPattern.java
package demo.java;

import java.io.FileInputStream;
import java.io.IOException;

public class LoanPattern {
  public static void main(String[] args) throws IOException {
    try(FileInputStream input = new FileInputStream("file.txt")) {
      int data = input.read();
      while(data != -1){
        System.out.print((char) data);
        data = input.read();
      }
    }
  }
}
